### ISSUE TITLE
Unable to enter DFU mode using a GPIO pin

### ISSUE THREAD

**Issue by tuzz** [gregdavill/OrangeCrab-examples#39]
Hello, I'm fairly new to FPGA development and I apologise if this is an obvious mistake.

I'm trying to get the Orangecrab (85F) to enter DFU mode when I press an external momentary switch that I have wired up on a breadboard. I have connected 3V3 from the Orangecrab through a switch and into GND via a 10k ohm resistor. I then adapted the `blink_reset.v` example to read `input gpio_10` and set `reset_sr` low when the button is pressed.

When I press the switch the orange crab resets. When I press and hold the switch, the onboard RGB LEDs are high (blue/green are maybe 50% and red is maybe 25%). In either case, the device doesn't enter DFU mode. The LEDs stop flashing during this time which leads me to think the device keeps resetting? I've tried debouncing the switch but that didn't seem to work. I'd appreciate any help you can provide. Thank you.

```verilog
// blink_reset.v - modified
module top (
    input clk48,

    output rgb_led0_r,
    output rgb_led0_g,
    output rgb_led0_b,

    output rst_n,
    input usr_btn,
    input gpio_10
);
    // Create a 27 bit register
    reg [26:0] counter = 0;

    // Every positive edge increment register by 1
    always @(posedge clk48) begin
        counter <= counter + 1;
    end

    // Output inverted values of counter onto LEDs
    assign rgb_led0_r = ~counter[24];
    assign rgb_led0_g = ~counter[25];
    assign rgb_led0_b = 1;

    // Reset logic on button press.
    // this will enter the bootloader
    reg reset_sr = 1'b1;
    always @(posedge clk48) begin
        reset_sr <= {~gpio_10};
    end
    assign rst_n = reset_sr;


endmodule
```

**Comment by gregdavill** [comment]
The bootloader in the orangecrab makes use of the ECP5's "multi-bitstream" feature. On POR the bootloader image is loaded. This checks the state of btn0, the button on the main board. If it's held then the bootloader starts up the USB and waits for an image. However if the button isn't pushed it reloads the FPGA into the users bitstream. 

From your logic you are likely correctly swapping back to the bootloader image, but the bootloader is then jumping back to your code because btn0 isn't pressed.


As a reference here is the bootloader code: https://github.com/gregdavill/foboot/tree/OrangeCrab

**Comment by tuzz** [comment]
I see, thank you. So does that mean there is no way to enter the bootloader from a GPIO pin signal, rather than using that exact physical button (`btn0`) on the board? I was hoping to remotely program and reprogram the Orangecrab for a project but that won't work if I have to be physically present to hold the button down. Thanks again.

**Comment by gregdavill** [comment]
With the standard bootloader that is correct. If you have a JTAG adapter supported by openFPGALoader (lots of FTDI based devices work) You can modify and update the bootloader with one that would let you switch back as you describe.

Possibly the signal ADC SENSE_LO can do this. User application charges it, and the bootloader can check the digital state?

**Comment by tuzz** [comment]
Thanks for the help, I will take a look and report back with my findings. I will close this now. Cheers.

**Comment by tuzz** [comment]
Just a quick update for posterity.

I managed to get this working so that I can (remotely) SSH in and program the Orangecrab using the GPIO pins on my Raspberry Pi and openFPGALoader. I soldered some header pins into the JTAG connectors on the Orangecrab and bent them into rows on my breadboard. I wrote [this script](https://gist.github.com/tuzz/4ffe7c187c5ed118f8e3ed397a3fc8b8) to bootstrap my Raspberry Pi which also synthesizes and programs the FPGA. The power/ground wires aren't necessary, I just wanted to distribute power so I can connect other peripherals.

![IMG_0899](https://github.com/user-attachments/assets/30542481-e52b-406e-84e3-a16aa8ac27fc)
